stake-pool/js/src/index.ts

```diff
@@ -254,15 +338,15 @@ export async function withdrawStake(
   // Check withdrawFrom balance
   if (tokenAccount.amount.toNumber() < poolAmount) {
     throw new Error(
-      `Not enough token balance to withdraw ${lamportsToSol(
-        poolAmount,
-      )} pool tokens.
-        Maximum withdraw amount is ${lamportsToSol(
-          tokenAccount.amount.toNumber(),
-        )} pool tokens.`,
+      `Not enough token balance to withdraw ${lamportsToSol(poolAmount)} pool tokens.
+        Maximum withdraw amount is ${lamportsToSol(tokenAccount.amount.toNumber())} pool tokens.`,
     );
   }
 
+  const stakeAccountRentExemption = await connection.getMinimumBalanceForRentExemption(
+    StakeProgram.space,
+  );
+
   const withdrawAuthority = await findWithdrawAuthorityProgramAddress(
```

Comment:
A few lines below this one, there is the following code:

```
if (useReserve) {
  withdrawAccounts.push({
    stakeAddress: stakePool.account.data.reserveStake,
    voteAddress: undefined,
    poolAmount,
  });
}
```

Won't this always fail if there is still stake in the validator accounts? In that case, I think it might be better to remove this part and pass `useReserve` into the `prepareWithdrawAccounts` method.

Then, in `prepareWithdrawAccounts`, we can change this:
```
if (reserveStakeBalance > 0) {
  accounts.push({
    type: 'reserve',
    stakeAddress: stakePool.reserveStake,
    lamports: reserveStakeBalance,
  });
}
```
to this:
```
if (useReserve && reserveStakeBalance > 0) {
  accounts.push({
    type: 'reserve',
    stakeAddress: stakePool.reserveStake,
    lamports: reserveStakeBalance,
  });
}
```

Reply:
> A few lines below this one, there is the following code:
> 
> ```
> if (useReserve) {
>   withdrawAccounts.push({
>     stakeAddress: stakePool.account.data.reserveStake,
>     voteAddress: undefined,
>     poolAmount,
>   });
> }
> ```
> 
> Won't this always fail if there is still stake in the validator accounts? In that case, I think it might be better to remove this part and pass `useReserve` into the `prepareWithdrawAccounts` method.
> 
> Then, in `prepareWithdrawAccounts`, we can change this:
> 
> ```
> if (reserveStakeBalance > 0) {
>   accounts.push({
>     type: 'reserve',
>     stakeAddress: stakePool.reserveStake,
>     lamports: reserveStakeBalance,
>   });
> }
> ```
> 
> to this:
> 
> ```
> if (useReserve && reserveStakeBalance > 0) {
>   accounts.push({
>     type: 'reserve',
>     stakeAddress: stakePool.reserveStake,
>     lamports: reserveStakeBalance,
>   });
> }
> ```

When processing withdrawals, the order of priority goes:

preferred withdraw validator stake account (if set)
validator stake accounts
transient stake accounts
reserve stake account

Reply:
Right, that's what it should be. However, when `useReserve` is set to `true`, the order becomes:

reserve stake account

which will never work because: `The lamports in the validator stake account is not equal to the minimum`

The stake pool program won't allow for a stake withdrawal from reserve stake account until all the validator stake accounts have been fully withdrawn.

Reply:
argh ... checking .. anyway to withdraw from reserve account you should use withdrawSol

Reply:
Right of course, `withdrawSol` is most optimal, `useReserve` seems a bit out of place as a parameter and can probably be removed altogether.

Reply:
I think I have a solution for the other bug, I'll add another comment once I test that it works (currently dealing with splitting transactions up since transaction sizes can get a bit large sometimes).